Lagoon alpha

sigilante opened the pull request:
# Lagoon for Urbit
## Linear Algebra in hOON

LAGOON, the _Linear AlGebra in hOON_ library, implements a suite of basic matrix operations for numerical applications. Lagoon endeavors to be mathematically accurate and notationally convenient. The fundamental problem Lagoon solves is representing memory-adjacent array data in a uniformly accessible way to Hoon.

What I do have done:
- definition of auras `@lvs`, `@lvd`, `@lms` (and their presumed analogues for other types)
- vector library + jets in `@rs` and `@rd`
- matrix library + jets in `@rs`

What I don't have done:
- matrix library + jets in `@rd`
- profiling data on performance
- some utility routines, in particular pretty-printing and import/export

I considered adding in ATLAS, TensorFlow, or another BLAS package as a dependency but ultimately decided not to for the alpha version.  Instead everything is done via SoftFloat as with other Urbit floating-point operations.  I have some basic unit test setups but as of Ford Fusion it seems that unit testing is itself broken.  Large tests and rigorous profiling will require import/export routines as well to avoid tedious manual entry.  I'd also like jet-by-jet comparison with the Hoon, which I didn't have time to finish.

The future roadmap:
- `@lmd` double-precision matrix type.  (Shouldn't be hard at all, just swapping a few bit widths.)
- Full unit testing suite compatible with whatever the recommended practice is or will be.
- More utility routines, firstly I/O (pretty-printing, file loading, and efficient CLI input).
- Prospectus for a sister library of scientific algorithms (based on `lazytrig` but jetted), provisionally Saloon (Scientific ALgorithms in hOON).
- Vectors and matrices in other significant types (`@ud`, `@rh`, `@rq` come to mind).  Creation of a complex type, without which certain operations (like eigenvalues/eigenvectors) will not be supportable.
- Code refactor of Lagoon:  I have wrought terrible things, like being inconsistent on 0-indexing and 1-indexing.  The code works but should be prettier, more consistent, and more legible.  Variable names should indicate their $n$-indexing if they differ.

Another question for consideration:
- Does atom order make sense?  As things stand, the Hoon is relatively straightforward but the C jet code keeps flipping arrays backwards and forwards to simplify the math.

sigilante commented:
Lagoon introduces a new aura @l in two flavors:

@lv denotes vectors, with s or d indicating @rs or @rd components.
@lm denotes matrices, with s indicating @rs componeents.
A vector atom is packed with the length pinned as the head (least significant byte) and the rest of the vector following bytewise in the atom.

A matrix atom is packed with the dimensions pinned as the head and the rest of the matrix following bytewise in the atom, row-major. The dimensions are packed into a single value using the Cantor pairing function, which unambiguously maps ordered pairs to integers.

As of the alpha release, Lagoon provides the following operations. All operations are jetted in C. There are currently no dependencies beyond SoftFloat.

For @lvs and @lvd:

++adds, scalar addition
++subs, scalar subtraction
++muls, scalar multiplication
++divs, scalar division
++addv, vector addition
++subv, vector subtraction
++mulv, vector multiplication (elementwise)
++divv, vector division (elementwise)
++zeros, zero-loaded vector
++ones, one-loaded vector
++length, vector length
++get, getter for index
++set, setter for index
++sum, sum of vector elements
++product, product of vector elements
++inner, inner product of two vectors

For @lms:

++cantor, Cantor pairing function
++decantor, inverse pairing function producing ordered pair
++zeros, zero-loaded matrix
++ones, one-loaded matrix
++id, identity matrix
++get, getter for index
++set, setter for index
++getr, getter for row index, returns @lvs
++setr, setter for row index, returns @lvs
++getc, getter for column index, returns @lvs
++setc, setter for column index, returns @lvs
++trans, transposition
++addsc, scalar addition to column index
++subsc, scalar subtraction to column index
++mulsc, scalar multiplication to column index
++divsc, scalar division to column index
++addsr, scalar addition to row index
++subsr, scalar subtraction to row index
++mulsr, scalar multiplication to row index
++divsr, scalar division to row index
++addvc, vector addition to column index
++subvc, vector subtraction to column index
++mulvc, vector multiplication to column index
++divvc, vector division to column index
++addvr, vector addition to row index
++subvr, vector subtraction to row index
++mulvr, vector multiplication to row index
++divvr, vector division to row index
++addm, matrix addition
++subm, matrix subtraction
++mulm, matrix multiplication (elementwise)
++divm, matrix division (elementwise)
++mmul, matrix multiplication (linear algebraic)
++mpow, matrix exponentiation
++minor, matrix minor
++submatrix, submatrix
++augment, augmented matrix
++invert, matrix inversion by Gauss--Jordan elimination

ohAitch commented:
Gotta ask: no higher-order tensors? Obviously not all operations are translatable into higher dimensions, but specializing the auras to `⍴⍴` of 1 and 2 feels kind of weird.

What does the @ta (`a-z0-9_.~-`) encoding look like for all of these?

sigilante commented:
Certainly one can imagine an $n$-dimensional `@lt` variant, or even backing into just an `@l` variant that encodes its own dimensionality.  There's a tension here:  build a library that provides that covers most use cases (vectors/matrices) or build a fully general library.  It may be better to stack atom ops instead of building out $n$-dim variants:  think of how Pandas and R use DataFrames as essentially collections of 1D Series.

This philosophical discursus from my notes touches on the question:

Is the row vector (1,1,1) a 1×3 vector or just a simple vector of length 3 (with no "thickness")? It's a bit of a trick question; [formally](https://en.wikipedia.org/wiki/Row_and_column_vectors), row vectors form a vector space _row space_ while column vectors form a corresponding dual _column space_. Since the underlying elements are drawn from the same field for row vectors, column vectors, and matrices, we can handwave a bit of the formalism and treat row vectors and column vectors as special cases of 2D matrices with one dimension being flat.

However, implementations in several numerical languages and packages disagree on how to best accomplish this:

+ C: an array may be defined as "one-dimensional" (an element is indexed once, `vector[1]`) or "two-dimensional with a thickness of one" (an element is indexed twice, `vector[0][1]`).
+ C++: the Standard Template Library provides a `vector` type but this is a fast array, not a linear algebraic construct; [Boost](https://www.boost.org/doc/libs/1_70_0/libs/numeric/ublas/doc/vector.html) provides `boost::numeric::ublas::vector<double> v(3)`, for example.
+ Python (NumPy): an `ndarray` of size `(1,1)` is distinct from an array of size `(1,)`.  [Ref](https://stats.stackexchange.com/questions/284995/are-1-dimensional-numpy-arrays-equivalent-to-vectors)
+ MATLAB: all values have dimension, even scalars being 1×1 arrays; thus any vector of length 3 has already thickness 1. _This is very convenient when working with standard dimensions._  [Ref](https://www.mathworks.com/help/matlab/math/multidimensional-arrays.html)
+ [Julia](https://en.wikibooks.org/wiki/Introducing_Julia/Arrays_and_tuples#Row_and_column_vectors): a `1x3 Array{Int64,1}` is distinct from a `3-element Array{Int64,2}`.

These differences arise from implementation details. It seems most sensible to hew to the implicit mathematical standard: to wit, a vector is a matrix with one dimension being one, not a "zero-thickness matrix". This implies that row vectors and column vectors should in fact differ from each other in the Lagoon implementation.

We assume that the majority of interest in a linear algebra library will be derived from classical vector–matrix operations in one and two dimensions, rather than tensor operations. Higher-dimensional arrays can proceed analogously from the lower ones, with the salient difference that we do _not_ treat vectors and matrices as already having dimension in those directions. There are thus some quirks involved in the treatment of arrays of three dimensions and higher.

sigilante commented:
The auras are packed `@rs` and `@rd` types.

For instance the five-element `@lv` vector corresponding to `~[.1 .2 .3 .4 .5]` looks like this:

```
0x540a000004080000040400000400000003f800000
```

ohAitch commented:
Right, the shape of a vector `1 2` is `2`, the shape of a matrix `[1 2;]` is `1 2`, they are different values; the shape of the shape of `1 2` is `1`, the shape of the shape of `[1 2;]` is `2`. (The shape of the shape of the shape, i.e. the shape of any scalar like 1 or 2, is empty-array; whose shape is in turn `0`)
The thing I'm less sure about is those ending up in the *type*, to the extent of having to keep track of separate `adds` `addv` `addm` calling conventions.

Pipe dream counterproposal, could use work: an `@rdv` type which
- does something clever with bit widths / the *last* byte
- such that a scalar `@rd` is a valid `@rdv` with 0 dimensions(trailing byte of 0)
- a vector e.g. `5 6` ends with the bytes `1 2`, to designate a 1-dimension-dimensional tensor whose row dimensions is that it has two elements. This is ofc checked against the actual byte count of floats up to this point
- this *still* limits us to two dimensions for single-precision floats hmm. I suppose that is a similar constraint like the current implementation's "vectors can be at most 255 floats long", and should be further extended(nan-packing? stuffing a backwards offset in the last 0-3 bytes if they're not enough?)

ohAitch commented:
Sorry yes to rephrase: what are your thoughts on pretty-printing? Obviously `0x140a00000` is a `@ux`(minus dots), not a `@l`, and for something to be a hoon aura it needs to have a bijective literal notation for using them in source code. (The pretty-printer giving invalid-hoon hex if you try to print an unimplemented aura notwithstanding)

ohAitch commented:
Like, strawman proposal on the "extend the floating point aura" side of things: `.~1._2.` for a vector of two, `.~1._2.__3._4.` for a two by two square, `.~1._` and `.~1.__` for a one-element and one-by-one element vector and matrix respectively.

sigilante commented:
I haven't touched pretty-printing at all yet.  Your strawman proposal seems satisfactory if cumbersome for large matrices; it would be convenient to have a legible line-breaking version too.

The reason I elected to pin the dimensions at the head and to break v/m apart was because Urbit strips vectors of zeroes:  essentially it was a spoof.  A meaning-free placeholder and then a dimensional suffix could work okay:  I'd have to play with it seriously to feel how it works.

The problem is knowing which entries are part of the dimensions and which are part of the array.  If you pinned the dimensionality to the front (to block zeroes) and the individual dimensions at the end, that works.  That frees you from implementing the v/m/t part of the aura, leaving `@ls` or `@rst`, for instance.

ohAitch commented:
I was thinking the last(most significant) byte is the number of dimensions, then immediately beforehand the lengths of all except one of the dimensions(in turn computed from the total length). So you can just for example `(cat 5 .3.14 1)` to get `.3.14_` or `(rap 5 .3.14 .4.13 1 ~)` for `.3.14_4.13`, with the vector length "2" implied by the atom being 2+1 words in length; or `(rap 5 .1 .2 .3 .4 .5 .6 ~[3 2])` to get `.1_2_3__4_5_6`, a `2`-dimensional array(matrix) with rows of length `3`. (Or alternately it could be `.1_2__3_4__5_6`, not sure whether omitting the major or minor dimension makes more sense here)

If the number of dimensions is zero, then you append a 0 byte, which is lost; but if your meta-dimensionality is 0, then what you have is a single scalar, and giving it equivalent representation to plain `@rd` is outright desirable.

For multiline, I suspect you'd want to reuse `\ <whitespace> /` from [text literals](https://github.com/urbit/urbit/blob/master/pkg/arvo/sys/hoon.hoon#L4069-L4070), giving
```
=/  a=@rds
  .1_2_3__\
  /4_5_6
```

For a cleaner look, you could also have `_` and `__` accept optional trailing whitespace in the same way `.` digit separators do, but this would necessitate a terminator in the `.1_2_3__` "not a vector, but rather a 1x3 matrix" usage - perhaps another dot.

ohAitch commented:
This does leave the question of what to do with `⍬` - the empty vector, which I am going to arbitrarily transcribe as `.-_` `.~-_` etc. Since just using the atom `1` with no leading floats would be confused for a scalar, I'd tentatively go for an `ff` but pattern(e.g. `0xff.ffff` for the `@rh` `.~~-_`), with the idea of reserving the byte 255 as "negative one dimension"; maybe there's something cleaner. (A 0x0 matrix, on the other hand, can be just `(cat n 0 2)` - and possibly we only need that)

In @l of course, just `` `@lvd`~`` or w/e suffices, because the first byte or first two bytes will be 0 precluding trying to read any of the remaining atom.

A word on alignement: insofar as indirect atom allocations get word-aligned or cache-aligned or w/e, the data-first representation can be loaded directly into AVX registers, reshaped back and forth between vector and matrix without shifting over the whole array(capacity permitting), and memcpy concatenated without having to update both the atom length /and/ the dimension. Conversely, the metadata being at the end does put an effectively pointer-indirection cost on *reading* it, possibly across a significant distance. Not sure how this shakes out in practice

ohAitch commented on pkg/arvo/lib/lv.hoon:
```diff
@@ -0,0 +1,546 @@
+      ?:  =(count n)  +(max-arg)
+      =/  cur-val  `@rs`(get u +(count))
+      ?:  (gth:rs cur-val max-val)  $(count +(count), max-arg count)
+    $(count +(count), max-arg max-arg)
```
No `@lvs` variant? Also, I'd maybe go for initializing max from the first candidate
```
=/  max  n
|-  ^+  max
?~  n  max
=?  max  (gte:rs (get u n) (get u max))  n
$(n (dec n))
```
Also also, should `.nan` count as min, max, or error here? (I think in current formulation it's max)

sigilante replied:
Also mixed @rd and @rs... yeesh.

sigilante replied:
```hoon
  ::
  ::    Return larger of two single-precision floats.
  ++  max-rs
    |=  [a=@rs b=@rs]  ^-  @rs
    ?:  (gth:rs a b)  a  b
  ::
  ::    Find maximum value in array.
  ++  max
    |=  [u=@lvs]  ^-  @rs
    ~_  leaf+"lagoon-fail"
    `@rs`(reel (unmake u) max-rs)
  ::
  ::    Return index of maximum value in array, 1-indexed
  ::      array.index(max(array)) is simple and fast
  ++  argmax
    |=  [u=@lvs]  ^-  @ud
    ~_  leaf+"lagoon-fail"
    (find ~[(max u)] u)
```

ohAitch commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    =/  t  (div (add (mul w w) w) 2)
+    =/  y  (sub z t)
+    =/  x  (sub w y)
+    ~[x y]
```
Honestly I'd just use `(rep 4 [m n]~)` here, cantor of two `@ud` into a `u32` gives you <2x range of just having a pair of u16, and outright less if your matrices are anywhere near square.

sigilante replied:
Adopted throughout.

ohAitch commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    =/  m   `@ud`-:mn
+    =/  n   `@ud`+<:mn
+    =/  size  (mul m n)
+    =/  p  (cantor m n)
```
unused

ohAitch commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    |=  [u=@lms]  ^-  (list (list @rs))
+    ~_  leaf+"lagoon-fail"
+    %-  flop
+    =/  mn  `(list @ud)`(shape u)
```
I feel like having `shape` return explicitly `[i=@ud t=[i=@ud t=~]]` or something would help while retaining compatibility, since then you could `=+  [m n ~]=(shape u)`

ohAitch commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+      ?:  (gth count m)  `@lvs`(make:lvs w)
+      =/  row  `(list @rs)`(snag (dec count) `(list (list @rs))`uu)
+      =/  elem  `@rs`(snag (dec i) row)
+    $(count +(count), w `(list @rs)`(weld w `(list @rs)`~[elem]))
```
I feel like a `makec/unmakec/maker/unmaker` that go <-> `(list @lv)` would help a lot with these

ohAitch commented:
I think I have figured out some of the confusion above: `@ux`(and the unpretty atom printer `@l` `@x` etc) renders hex in LSB order, while memory order is LSB.

```
> `@x`%abc-c
0x632d636261
> `@x`'\01\02\03'
0x30201
```

Thus when you say
> a vector atom is packed with the length pinned as the head (least significant byte)

while you are right that it is printed out at the beginning of `@x`, the bytes those are the _most_ significant ones. This can be seen in the equivalency between `zeroed:lvs` and a left-shift:

```
> (zeros:lvs:lv 1)
0x100000000
> `@x`(lsh 5 1 1)
0x100000000
> (zeros:lvs:lv 3)
0x3000000000000000000000000
> `@x`(lsh 5 3 3)
0x3000000000000000000000000
```

(which in turn means the former doesn't really warrant its own jet - unless something is weird with the allocator)

My dimensioning scheme is to further prepend a dimension-length field, but leave the last dimension implied by the buffer length `(met 5 a)`; for vectors, this just replaces the length field with a buffer-end marker `1`.

Looks like the floats _themselves_ are in LSB order: that is, a hypothetical `.~~6_7_8` `(make:lvh (turn (gulf 6 8) sun:rh))` would be `0x3.4800.4700.4600` and not `0x3.4600.4700.4800`.

Not sure why
> the C jet code keeps flipping arrays backwards and forwards to simplify the math.

Is this still true in the current revision 54651ed? I see one comment in the matrix invert routine, next to what appears to be commented out inversion code.

ohAitch commented on pkg/arvo/lib/lv.hoon:
```diff
@@ -0,0 +1,546 @@
+    =/  sum  .0
+    |-  ^-  @rs
+      ?:  =(count n)  `@rs`sum
+    $(count +(count), sum (mul:rs sum (get u +(count))))
```
`(roll (unmake u) |=([a=@rs b=_.1] (mul:rs a b)))`

Filed #3624

ohAitch reviewed (commented):
First pass, mostly code style / simplification separate from the "why".

Haven't looked at jets or most of matrix code yet

ohAitch commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    =/  mnu  `(list @ud)`(shape u)
+    =/  mu  `@ud`-:mnu
+    =/  nu  `@ud`+<:mnu
+    ?.  =(mu nu)  !!  :: make sure this is a valid operation (square matrix)
```
```suggestion
    ?>  =(mu nu)  :: make sure this is a valid operation (square matrix)
```
there's a rune for that :)

ohAitch commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    =/  u  (augment u)
+    |-  ^-  @lms
+      ?:  (gth i mu)  (gauss-row-replace u)
+    $(i +(i), u (gauss-row-reduce u i))
```
```
++  gauss-row-replace
  |=  u=@lms  ^-  @lms
  =+  [m n]=[&1 &2]:(shape u)
  %+  roll  (gulf 1 m)
  |:([i=*@ u=(augment u)] (gauss-row-reduce u i))
::
++  gauss-elim
  |=  u=@lms  ^-  @lms
  =/  n  +:(mate [`&1 `&2]:(shape u))  :: make sure this is a valid operation (square matrix)
  %+  roll  (gulf 1 n)
  |:([i=*@ u=(augment u)] (gauss-row-reduce u i))
```
not sure the rolls in particular are actually less cluttered; I suppose at minimum they're differently cluttered?

ohAitch commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+      ?:  (gth jj nu)  $(ii +(ii), jj 1)
+      ?:  =(ii i)      $(ii +(ii))
+      ?:  =(jj j)      $(jj +(jj))
+      =/  iii  ?:((gth ii i) (dec ii) ii)
+      =/  jjj  ?:((gth jj j) (dec jj) jj)
+      $(jj +(jj), w (set w iii jjj (get u ii jj)))
```
This looks like it's cutting out a cross-shaped chunk?
```
=.  u  (maker (oust [(dec i) 1] (unmaker u)))
=.  u  (makec (oust [(dec j) 1] (unmakec u)))
u
```

ohAitch reviewed (commented):
Part 2/3: matrix hoon. Bold ideas on how to shorten things considerably. (And/or bring them closer to the bits.)

ohAitch commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    ~/  %invert
+    |=  [u=@lms]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    (gauss-elim u)
```
```suggestion  
    =+  [m n]=[&1 &2]:(shape u)
    (submatrix [1 m] [+(n) (mul 2 n)] (gauss-elim u))
```
per the jet. Otherwise you get the whole augmented matrix
```
> =,(lms.lm =/(a (make ~[.0.5 .0] ~[.0 .0.5] ~) (unmake (invert a))))
~[
  ~[.1 .0 .2 .0]
  ~[.0 .1 .0 .2]
]
```

sigilante replied:
:)

sigilante commented on pkg/arvo/lib/lv.hoon:
```diff
@@ -0,0 +1,546 @@
+    :: collides with this, but it's like getting a hash table collision, very
+    :: very unlikely.  Also the Urbit NaN is hard-coded, `SINGNAN`.
+    =/  nan  0xffff.ffff
+    `@lvs`(not 5 +(n) (rap 5 ~[(fil 5 n nan) (not 5 1 n)]))
```
Yes.  This sort of refactoring is what is needed so it doesn't feel so cobbled-together.

Pending the actual atom form, of course, for which we need more weigh-ins.

sigilante replied:
This yields the desired behavior for the current atom shape.  I've added it locally (will pull them in as a block).  I'll use :rocket: to indicate integration on each of these.

ohAitch replied:
Yeah, was definitely aiming for "preserve semantics" here. Though more generally the atom format is "row-major with some trailer words", swapping in half-word concat for cantor or 0x1 for (lent) or w/e would be fairly straightforward - idk if anyone else has opinions on the atom form.

More food for thought: ideally `.1_0j1` would be the complex basis vector…

sigilante replied:
I'm a little bugged that `@c` and `@i` are both occupied already for simple complex atoms...

ohAitch replied:
Yeah - otoh we could stretch the mnemonic of "real" somewhat and have `@rsr` `@rdr` etc be plain floats, and `@rs` be "LSB real then complex parts", since after all casting a 32-bit float to a pair of a 32-bit real and 32 trailing zero bits is perfectly valid.

ohAitch replied:
Then in my representation, use e.g. the top bit of the shape-length byte for "is complex", so `0x1` is the "vector of reals" marker and `0x81` is the "vector of complex" marker - means if you want transparent scalars you can't have a vector of 1 real, just a vector of 1 complex, which is kinda inefficient, but not sure it's a huge deal. 
(Alternately do something clever with NaNs and a non-word-aligned trailing byte)

Or you could give up on untagged scalars and full-on have both the shape and complexity and while we're at it bitwidth in the trailer, producing a new `@rt` "float tagged / float tensor" format. Would probably want additional literal syntax for "one float/complex, not in a vector or matrix" and generally make the tensor one less "just a float but plural" - is `..` a used prefix?

ohAitch replied:
In general this gets you quite an aura stack, but I think they all telescope properly:
```
@rt - float tagged
@rts - float tagged single-precision
@rtsu - float tagged single-precision unitary(singular)
@rtsur - float tagged single-precision one real
@rtsuc - float tagged single-precision one complex
@rtsvr - float tagged single-precision vector of real
@rtsmc - float tagged single-precision matrix of complex
(@rtr - float tagged some-precision of some-shape of reals, etc)
```

ohAitch replied:
#3759 up with `AjB` parsing/printing - haven't done any of the math support, kinda want to pull it out into zuse and leave hoon.hoon with just scalar mul/div/round/dragon4 as necessary for textual serdes.

sigilante replied:
Following up:  this implies that the primitive complex value is actually a 1×1 complex matrix or a 1× complex vector.  Should there also be a "bare" complex aura?  Do we back into a corner by only having in-structure complexes?

sigilante replied:
Let me write up a proposal with several possibilities and submit it as a separate issue.

sigilante replied:
I've added my four cases to #3759 now for discussion.  Pending that, I'll start resolving the atom structure for `@l` or `@rt`.

ohAitch replied:
The main distinction in my mind is "are @rd bit-compatible with lagoon scalars", sort of in the same way that `@ux` can be a "view" onto `@t` but even more tightly corresponding.

If we go the route of "rank bitwidth etc live in the auras"(and only the auras), then `@lus` would be a "lagoon scalar", which is different from `@rs` only in that the trailing imaginary halfword can be nonzero. This gives you a cleaner separation between tensor/complex-capable and tensor-incapable code, with speedup from avoiding dynamic dispatch, at the cost of having to care about that this distinction in all calling code. (inserting various typecasts, duplicating interfaces for all combinations, doing the same if the calling code is itself a general-purpose library.)

If we go the route of "rank bitwidth etc are encoded in the value"(i.e. the trailing byte scheme), then generic `@rt` is usable everywhere, and the more specific auras are only relevant for constants / when you are sure you want only matrices or only reals or w/e. (Though as more of a "flagset" than conventional aura nesting, we'd either want only some of the tree expressly encoded, or to change compiler nesting semantics to be more filter-list like.) This gives you an easier to follow surface area, but requires runtime conversion to interact with existing float code ~~which there isn't any of~~, and of course if you *do* type everything as just `@rt`, you might end up with a complex vector where you wanted a real matrix. Library functions like `add` in particular will want to be superficially-wet, in that there is a core implementation that works on all `@rt`(but checks size and probably-format compatibility), and then a wrapper around it that calculates the result aura from that of `a` and `b`(and type-fails on statically known mismatch).

There is also a possibility of e.g. encoding the rank but not the individual float `bloq`, which I think was my earlier proposal, but it seems likely correct to go all-in one way or the other. And given the *shape* is not in the aura anyway, I'm leaning towards "~all\~ the runtime type information".

ohAitch commented on pkg/urbit/jets/e/lms.c:
```diff
@@ -0,0 +1,2134 @@
+    {
+      for ( jj = 0; jj < n_u; jj++ )
+      {
+        w_[ii*n_w+n_u+jj] = u3r_word(ii*n_u+jj, u);
```
This `+n_u`, granted, would need to move to the next assignment if the word order was swapped from MSB to LSB.

ohAitch commented on pkg/urbit/jets/e/lms.c:
```diff
@@ -0,0 +1,2134 @@
+  }
+
+/* invert matrix
+   XXX this should become a wrapper not an implementation
```
wrapper to sth like BLAS?

sigilante replied:
Right now I've avoided introducing any external dependencies besides SoftFloat.  I looked into doing more but I think the platform and installation situation starts to get more complicated and I'd like more dev weigh-in before taking the step of increasing the number of dependencies.

sigilante replied:
What I think makes the most sense is to have a base matrix/vector library (this one), then a scientific algorithms library (everything past inversion).  Once you get past inversion (eigenvalues and so forth), there are a lot of interdependencies introduced very quickly and the size will balloon.  So these basic data structures become standard, but the scientific/analytical stuff only gets pulled in when necessary.

ohAitch replied:
Yeah makes sense, this tbc was a "what does the comment mean" comment not a "so do that" comment :P

ohAitch reviewed (commented):
3/3: Jets!
Require extensive hardening for weirdly sized inputs; don't have any obvious memory leaks or anything. Not sure what's going on with the matrix-specific word order reversal.

lvd.c and lvs.c are largely identical so most comments apply across both :P (similarly `adds` vs `subs` vs `muls` largely vs `addv` vs …)

sigilante commented on pkg/arvo/lib/lv.hoon:
```diff
@@ -0,0 +1,546 @@
+    ::  we have to pin values on both ends to compensate, then trim.
+    =/  nan  0xffff.ffff
+    =/  ut  (swp 5 (cat 5 nan (cat 5 u nan)))
+    (oust [n 1] (oust [0 2] (rip 5 ut)))
```
It had to do with handling leading zeroes (or trailing zeroes once flipped).  This version avoids that issue.

sigilante commented on pkg/arvo/lib/lv.hoon:
```diff
@@ -0,0 +1,546 @@
+    |=  [u=@lvs i=@ud]  ^-  @rs
+    ~_  leaf+"lagoon-fail"
+    =/  n  `@ud`(length u)
+    (snag (dec i) (unmake u))
```
Indexing consistency was a bit of a mess, and in particular in the jet code I need to indicate 0 v. 1 indexing.

sigilante replied:
`(cut 5 [(dec i) 1] u)`

sigilante commented on pkg/arvo/lib/lv.hoon:
```diff
@@ -0,0 +1,546 @@
+    (snag (dec i) (unmake u))
+  ::
+  ::    Pretty-print the contents of the vector.
+  ++  pprint  !!
```
That works for starters (as long as I remember the `tape`...)

sigilante commented on pkg/arvo/lib/lv.hoon:
```diff
@@ -0,0 +1,546 @@
+  ++  length
+    |=  u=@lvs  ^-  @ud
+    ~_  leaf+"lagoon-fail"
+    (snag 0 (flop (rip 5 u)))
```
Went with `(end 5 1 (swp 5 u))`

sigilante commented on pkg/arvo/lib/lv.hoon:
```diff
@@ -0,0 +1,546 @@
+    ~/  %set
+    |=  [u=@lvs i=@ud x=@rs]  ^-  @lvs
+    ~_  leaf+"lagoon-fail"
+    (make (weld (scag (dec i) (unmake u)) [x (slag i (unmake u))]))
```
Yes, I have been inconsistent on bounds checking (as opposed to sometimes crashing).

Build it with `cat` and `cut` directly, perhaps.

sigilante replied:
`(cat 5 (cat 5 (cut 5 [0 i] u) x) (cut 5 [+(i) (length u)] u))` ought to work

sigilante replied:
I left a memo in there to switch to `++sew` at some point but the current code works I think.

sigilante commented on pkg/arvo/lib/lv.hoon:
```diff
@@ -0,0 +1,546 @@
+    =/  w  `(list @rs)`~
+    =/  count  0
+    |-  ^-  @lvs
+      ?:  =(count n)  (make w)
+      =/  ul  (snag count uu)
+      =/  wl  `@rs`(add:rs ul v)
```
I'm going to simplify all of the scalar arithmetic to the following pattern:

```
    =/  ss  (fill (length u) v)
    (addv u ss)
```

and do the fancier work in the vector arithmetic arms.

sigilante commented on pkg/arvo/lib/lv.hoon:
```diff
@@ -0,0 +1,546 @@
+      ?:  =(count n)  (make w)
+      =/  ul  (snag count uu)
+      =/  wl  `@rs`(add:rs ul v)
+    $(count +(count), w (weld ~[wl] w))
```
Ah, I see, you suggest using `++turn` instead.  Let me do that in the vector arithmetic arms per your enhanced `++turnv` suggestion below.

sigilante commented on pkg/arvo/lib/lv.hoon:
```diff
@@ -0,0 +1,546 @@
+      =/  ul  (snag count uu)
+      =/  vl  (snag count vv)
+      =/  wl  `@rs`(add:rs ul vl)
+    $(count +(count), w (weld ~[wl] w))
```
Good stuff.

sigilante commented on pkg/arvo/lib/lv.hoon:
```diff
@@ -0,0 +1,546 @@
+    =/  sum  .0
+    |-  ^-  @rs
+      ?:  =(count n)  `@rs`sum
+    $(count +(count), sum (add:rs sum (mul:rs (get u +(count)) (get v +(count)))))
```
These weren't written in order so there was re-re-re-duplication of effort sometimes.

sigilante replied:
Yeah, you could do it with `fma:rs` but the sum expression is fine too and a bit less obscure.

sigilante commented:
As of this morning, I have integrated all `lv.hoon` suggestions for both `lvs` and `lvd` and pushed to the branch.  Spot checks indicate that things are working properly, and @ohAitch's generous commentary led to a felicitous decrease in the line count of about 200 lines of code.

sigilante commented on pkg/urbit/jets/e/lvs.c:
```diff
@@ -0,0 +1,562 @@
+  union trip {
+    float32_t s;    //struct containing v, uint_32
+    c3_w c;         //uint_32
+    float b;       //float_32
```
`trip` was really really useful for debugging `printf`s.  I'd keep it around, even if quiet, just for that, but move it out of functions generally.

ohAitch replied:
Personally I'd leave that as a commented-out variant `// float b //useful for debugging printfs`, but if float32_t and c3_w just directly cast to each other, sure maybe just do that.

ohAitch commented:
> Vectors and matrices in other significant types (@ud, @rh, @rq come to mind)

So for `@ud`, except for the issue of trailing zeroes `@ub` is effectively a vector of bits(`@f`), and `@ux` is a vector of nybbles or bytes or halfwords depending how you want to look at it. In a tagged `@rt` regime, I imagine there could also be a `@ut` that similarly uses the trailing byte to store its shape and bitwidth/signedness (probably not much of a call for complex integers), and some bits immediately preceding the shape trailer to disambiguate sub-byte length. Similarly we already have char arrays, though perhaps not `simdjson` style char matrices…

ohAitch commented:
Oh hey this PR exists! On the implementation slice, one thing to note is #4049 (in particular `run` and `sew`) should simplify a lot of the atom logic.

joemfb commented:
@ohAitch, they were added with this PR (and your prodding) in mind.

@sigilante, we changed some low-level atom gates in hoon %140 -- `+rep` absorbed `+repn` and `+rip` absorbed `+ripn`; while `+end`, `+lsh` and `+rsh` were changed to match the `+rep`/`+rip` API. And new additions `+run`, `+rut`, and `+sew` are likely relevant (although not yet directly jetted). 

Those, plus the jet declaration restaging that went with hoon %141 -> %140, make for a bunch of breaking changes with this PR. I'd be happy to help with the merge, as I've created the work. I'll look into making a PR to your fork.

sigilante commented:
@joemfb I'm going to get this done before the semester ramps up, so I'll read the new docs on `rep` etc. and weigh in again early next week.

ohAitch commented:
The short version is `(rep 3 x)` can now be written either `(rep 3 x)` or `(rep [3 1] x)`, `(repn 15 x)` is now `(rep [0 15] x)` and this generalizes to `(end 3 1 y)` turning into `(end 3 y)`/`(end [3 1] y)` etc for the other functions in the PR.

sigilante commented:
OK, `@lvs` and `@lvd` are updated under the current structure, the variable names have been refactored for consistency, and an off-by-one error was corrected.

sigilante commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    |=  [u=(list (list @rs))]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  m  `@ud`(lent u)
+    =/  n  `@ud`(lent (snag 0 u))
```
Controversial:  shouldn't a 0-D matrix just be a vector?  NumPy and MATLAB, for instance, differ on this point:

+ C: an array may be defined as "one-dimensional" (an element is indexed once, `vector[1]`) or "two-dimensional with a thickness of one" (an element is indexed twice, `vector[0][1]`).
+ C++: the Standard Template Library provides a `vector` type but this is a fast array, not a linear algebraic construct; [Boost](https://www.boost.org/doc/libs/1_70_0/libs/numeric/ublas/doc/vector.html) provides `boost::numeric::ublas::vector<double> v(3)`, for example.
+ Python (NumPy): an `ndarray` of size `(1,1)` is distinct from an array of size `(1,)`.  [Ref](https://stats.stackexchange.com/questions/284995/are-1-dimensional-numpy-arrays-equivalent-to-vectors)
+ MATLAB: all values have dimension, even scalars being 1×1 arrays; thus any vector of length 3 has already thickness 1. _This is very convenient when working with standard dimensions._  [Ref](https://www.mathworks.com/help/matlab/math/multidimensional-arrays.html)
+ [Julia](https://en.wikibooks.org/wiki/Introducing_Julia/Arrays_and_tuples#Row_and_column_vectors): a `1x3 Array{Int64,1}` is distinct from a `3-element Array{Int64,2}`.

sigilante replied:
Gonna punt on this for future philosophers.

joemfb commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    |=  [u=@lms i=@ud j=@ud x=@rs]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    `@lms`(setr u i (set:lvs (getr u i) j x))
+  ::
```
Added (with a slightly different API) in the breach:

```
> `@ux`(sew 2 [1 3 0xff] 0x1234.5678)
0x1234.0ff8
```

sigilante replied:
In the cell argument:  first value is the zero-indexed position, third is the fill.  What is the second number doing?

sigilante commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    =/  i  1
+    |-
+      ?:  |(=(i m) =(i n))  u
+    $(i +(i), u (set u i i .1))  :: TODO test Hoon version (jet works)
```
Changed to one arg; was two because of MATLAB syntax which allows asymmetric "identity" matrices.

`++repn` changed so this version doesn't work; I can munge the bits later but I'm marking this done for now.

sigilante commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+      =/  ell  `@rs`(get:lvs c count)
+      =/  trg  (set:lvs (make:lvs src) j ell)
+      =/  row  (unmake:lvs trg)
+    $(count +(count), w `(list (list @rs))`(weld w `(list (list @rs))`~[row]))
```
This has been obviated by the bitwise work done since.

sigilante commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    |=  [u=@lms i=@ud j=@ud]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    =/  t  (getc u j)
+    (setc (setc u j (getc u i)) i t)  :: TODO I think this Hoon may be broken
```
(The break was upstream in `++getc`.)

sigilante commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+      ?:  (gth count n)  `@lms`w
+    $(count +(count), w (setr w count (getc u count)))
+  ::
+  ::  Arithmetic operators
```
Opted for lv-style `turn2` instead of making/unmaking.

sigilante commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    |=  [u=@lms i=@ud v=@lvs]  ^-  @lms
+    ~_  leaf+"lagoon-fail"
+    (setr u i (divv:lvs (getr u i) v))
+  ::
```
I think this is too heavy for the basic matrix ops in here, but could make its way into a supplementary library with something more like actual numerics.

sigilante commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    =/  count  1
+    |-  ^-  @lms
+      ?:  (gth count mu)  `@lms`w
+    $(count +(count), w (mmul u w))  :: TODO check this in Hoon
```
beeyooteefull

sigilante commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    =/  count  1
+    |-  ^-  @rs
+      ?:  (gth count mu)  sum
+    $(count +(count), sum (add:rs sum (get u count count)))
```
Done.

sigilante commented on pkg/arvo/lib/lm.hoon:
```diff
@@ -0,0 +1,702 @@
+    $(count +(count), sum (add:rs sum (get u count count)))
+  ::
+  ::    Operations related to matrix inversion
+  ::    As with matrix multiplication, we're opting for clarity, not efficiency.
```
Is there a ++catenate:lvs I missed somewhere?

ohAitch replied:
It's just like `|=([a=@lvs b=@lvs] =/(c (length a) (mix (end 5 c a) (lsh 5 c b))))` I think?
Or `|=([a=@lvs b=@lvs] (make (weld (unmake a) (unmake b))))` which is much slower but perhaps clearer.

ohAitch replied:
Oh right implicit length isn't in the current `@l` format, yeah just use the latter with perhaps a jet :P

sigilante commented:
Current status:

- rewritten lv/lm still in `@l` aura type
- refactored all Hoon code to use consistent conventions
- waiting on `++catenate:lvs` clarification
- need to convert ad hoc tests to full test suite in `-ted`
- will start on updating and refactoring C jets

sigilante commented:
All right, the test suite is updated and all of the Hoon syntax changes have been fixed.

Next steps:
- [ ] `@lmd` double-precision matrix version (quick)
- [ ] C jets

So there's a light starting to glimmer at the end of this!

sigilante commented:
> Those, plus the jet declaration restaging that went with hoon %141 -> %140, make for a bunch of breaking changes with this PR. I'd be happy to help with the merge, as I've created the work. I'll look into making a PR to your fork.

I'm getting a lot of errors of the type

```
fund: in lvs, parent bfa3f13 not found at 7
fund: in get, parent 7b61e40f not found at 7
```

Comparing to the updated changes in %140,

- Should I use `~%  %lvs  +  ~` or `~%  %lvs  +>  ~` to start off the jet tree now?
- Do I need to version-pin it like `~%  %k.140  ~  ~`?
- What is `+|` doing in other files like `hoon.hoon`?

sigilante commented:
I think I've gotten things working with `..part`—at least, the warnings went away!

sigilante commented:
I'm breaking this up because it's become hard to track what's done and what's not.

#4455 has the Hoon vector code.

sigilante commented:
It occurred to me in connexion with the matrix library that at least cubic arrays should receive first-class support because of image processing.

tacryt-socryp commented:
How's this going?

sigilante commented:
For my part, it's mostly refactoring existing code and getting a few more unit tests in place for the matrix work.  (those are probably 95% done tho)

I should have time to get a good run in on it next week now that the semester is over.
